src/com/amazon/ion/impl/IonReaderBinaryIncremental.java

```diff
@@ -1245,13 +1249,14 @@ public SymbolTable getSymbolTable() {
 
     @Override
     public SymbolTable pop_passed_symbol_table() {
-        if (hasTransferredSymbolTable) {
+        SymbolTable currentSymbolTable = getSymbolTable();
+        if (currentSymbolTable == symbolTableLastTransferred) {
             // This symbol table has already been returned. Since the contract is that it is a "pop", it should not
             // be returned twice.
             return null;
         }
-        hasTransferredSymbolTable = true;
-        return getSymbolTable();
+        symbolTableLastTransferred = currentSymbolTable;
+        return symbolTableLastTransferred;
```

Comment:
The previous logic worked as long as `pop_passed_symbol_table` wasn't called more than once with the reader positioned at the same value. While making these changes I learned that there is logic in the DOM that does in fact do that.